TrustKit/TrustKit.m

```diff
@@ -9,130 +9,146 @@
-                                                                                    kTSKValidationCertificateChainNotificationKey: certificateChain,
-                                                                                    kTSKValidationNotedHostnameNotificationKey: notedHostname,
-                                                                                    kTSKValidationServerHostnameNotificationKey: serverHostname}];
-                   });
+    self = [super init];
+    if (self && [trustKitConfig count] > 0) {
+        // Convert and store the SSL pins in our global variable
+        _configuration = parseTrustKitConfiguration(trustKitConfig);
+        
+        _validationDelegateQueue = dispatch_get_main_queue();
+        
+        // Create a dispatch queue for activating the reporter
+        // We use a serial queue targetting the global default queue in order to ensure reports are sent one by one
+        // even when a lot of pin failures are occuring, instead of spamming the global queue with events to process
+        _pinFailureReporterQueue = dispatch_queue_create(kTSKPinFailureReporterQueueLabel, DISPATCH_QUEUE_SERIAL);
+        
+        // Create our reporter for sending pin validation failures; do this before hooking NSURLSession so we don't hook ourselves
+        _pinFailureReporter = [[TSKBackgroundReporter alloc] initAndRateLimitReports:YES];
+        
+        // Configure the pinning validator and register for pinning callbacks in order to
+        // trigger reports on the pinning failure reporter background queue.
+#if TARGET_OS_IPHONE
+        BOOL userTrustAnchorBypass = NO;
+#else
+        BOOL userTrustAnchorBypass = [_configuration[kTSKIgnorePinningForUserDefinedTrustAnchors] boolValue];
+#endif
+
+        static dispatch_once_t onceToken;
+        dispatch_once(&onceToken, ^{
+            sharedHashCache = [[TSKSPKIHashCache alloc] initWithIdentifier:@"trustkit"];
+        });
+        
+        __weak typeof(self) weakSelf = self;
+        _pinningValidator = [[TSKPinningValidator alloc] initWithPinnedDomainConfig:_configuration
+                                                                          hashCache:[[TSKSPKIHashCache alloc] initWithIdentifier:kTSKSPKISharedHashCacheIdentifier]
+                                                      ignorePinsForUserTrustAnchors:userTrustAnchorBypass
+                                                              validationResultQueue:_pinFailureReporterQueue
+                                                            validationResultHandler:^(TSKPinningValidatorResult * _Nonnull result) {
+                                                            
+                                                                // Invoke client handler if set
+                                                                void(^callback)(TSKPinningValidatorResult *) = self.validationDelegateCallback;
```

Comment:
It will work as intended. The callback is not captured in the block, only the "self" reference, which is a pointer to a structure that has the `validationDelegateCallback` pointer in it.

There is a different issue that I should have noticed – the `self` here should be `weakSelf` to prevent retain cycles!